### DISCUSSION TITLE
struggeling with meta_predicate

### DISCUSSION THREAD

**Discussion by hnbeck** [LogtalkDotOrg/logtalk3#143] [Q&A]
Hi, 

although we had already a discussion about that ( #123 ) I'll still got not theses things with meta_predicate, neither from SWI Prolog side, nor Logtalk side. And I have read both documentations :) Modules and there consequences seems to be hard to understand...

Well, my concrete problem.

See
```logtalk
:- object(factRules).
	:- public([	ruleSetup/2, compileRule/2, updateFact/4 ]).
	:- uses(list, [append/3, length/2, member/2, select/3]).
	:- dynamic(fact/2).
	%:- meta_predicate(factRules:functionValues, functionValues(*,*)).
```

Thats the beginning of an object. The predicate functionValues is

```logtalk
functionValues(Name, Pairs) :-
	zonelog::logMsg(detail, 'calculate Domain of ~w~n', [Name]),
	domain(Name, Domain),
	zonelog::logMsg(detail, 'Domain of ~w: ~w~n', [Name, Domain]),
	call(Name, Domain, [], Pairs).
```
inside this object. 
Then Logtalk complains: 

```text
*  Missing meta_predicate directive for predicate: functionValues/2
*       while compiling object factRules
*       in file c:/dev/psm_logic/facts/factrule.lgt between lines 168-172
```

Its Win10, Logtalk 3.54, SWI 8.2.4.  If I uncomment the meta_predicate above, there is the error

```text
!     Domain error: value (meta_predicate)/2 is not in domain directive
!       in directive meta_predicate/2
!       in file c:/dev/psm_logic/facts/factrule.lgt at or above line 9
```
I'm lost ..... ;)

Cheers

Hans

**Comment by pmoura** [comment]
The `meta_predicate/1` directive only have a single argument. But you're writing:

```logtalk
:- meta_predicate(factRules:functionValues, functionValues(*, *)).
```

That's why the Logtalk compiler complains and writes:

```text
!     Domain error: value (meta_predicate)/2 is not in domain directive
```

Assuming that the `functionValues/2` predicate is defined inside the `factRules` object, then it should be simply:

```logtalk
:- meta_predicate(functionValues(*, *)).
```

This directive assumes that `Name` is for a local object predicate. If that's not the case and `Name` is for a predicate defined in the sender of a `functionValues/2` message, then you would need instead:

```logtalk
:- meta_predicate(functionValues(3, *)).
```

But I don't see a public scope directive for the `functionValues/2` predicate and thus I assume you want the former directive. The reason why the compiler prints a warning if you comment out the directive, is to ensure that the compilation of the `call/4` goal is what indeed what the user intends.

Some additional observations:

If the clauses for the `fact/2` dynamic predicate are local to the `factRules` object, then add a private scope directive for the predicate for better performance:

```logtalk
:- private(fact/2).
:- dynamic(fact/2).
```

The last two arguments of the `zonelog::logMsg/3` predicate seem to be arguments for an implicit call to the `format/2-3` predicates. I suggest you consider using the [message-printing mechanism](https://logtalk.org/manuals/userman/printing.html) and possibly also [meta-massages](https://logtalk.org/manuals/userman/debugging.html#meta-messages) as an alternative.

**Reply by hnbeck** [reply]
Hi Paulo, 

thanks for you help and comments. Problem solved .... :)

But I would like to understand why functionValues/2 requires a meta_predicate directive. It is a custom predicate, not related to any SWI Prolog module - and as far as I understood from the Logtalk handbook, such conflicts arise only if module related code from SWI Prolog occurs .... I assume here I missing something about modules in SWI Prolog or SWI Prolog itself...

Cheers

Hans

**Reply by pmoura** [reply]
The `functionValues/2` predicate,  due to making a local meta-call, doesn't strictly require a `meta_predicate/1` directive. Note that the compiler prints a **warning**, not an **error**. The warning exists as technically the predicate is a meta-predicate. By adding a `meta_predicate/1` directive, that becomes more clear and better documented. The existence of the directive is also useful for code analysis tools. Finally, nothing here is related to module systems as we're not trying to call a Prolog module meta-predicate, which can have ambiguous directives.

**Reply by hnbeck** [reply]
Ah ok, now it's a little bit clearer, thanks :)

Cheers

Hans
